### ISSUE TITLE
EarlyStopping in the middle of an epoch

### ISSUE THREAD

**Issue by Richienb** [Lightning-AI/pytorch-lightning#19563]
### Description & Motivation

I'm fitting a normalizing flow to learn the mapping between two embedding spaces. The first embedding space is sampled using the mapper of a pretrained stylegan and the second embedding space is derived by a pretrained covnet. I want to learn a mapper from the second embedding space back to the first one. Since the stylegan can produce infinite data, I'm using an iterable dataset across one single epoch that encompasses the entire training run. So, I want `EarlyStopping` to trigger in the middle of the epoch. Validation data isn't available.

### Pitch

An option called `check_interval` should be added to `EarlyStopping`. If the value is a float, it is the fraction of an epoch between checks. If the value is an integer, it is the amount of training steps between checks. For the change to be non-breaking, its default should be `1.0`.

### Alternatives

Currently, I'm passing the EarlyStopping callback to the LightningModule and manually calling the check at the end of each training batch:

```py
def on_train_batch_end(self, outputs, batch, batch_idx):
    self.early_stopping_callback._run_early_stopping_check(self.trainer)
```
